File: RightViewOfTree.java
//TC : o(n)
//Sc: o(n)
//https://leetcode.com/problems/binary-tree-right-side-view/
//BFS solution
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        
        Queue<TreeNode> q = new LinkedList<>();
        List<Integer> result = new ArrayList<>();
        if(root == null) return result;
        q.add(root);
        while(!q.isEmpty()){
            int size = q.size();
            for(int i =0; i<size; i++){
                TreeNode curr = q.poll();
                if(i==size-1){
                    result.add(curr.val);
                }
                if(curr.left != null) q.add(curr.left);
                if(curr.right != null) q.add(curr.right);
            }
        }
        return result;
    }
}

//dfs on right

class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> result = new ArrayList<Integer>();
        dfs(root, 0, result);
        return result;
    }
    private void dfs(TreeNode root, int level, List<Integer> result){
        //base
        if(root == null) return;
        
        //logic
        if(result.size() == level){
            result.add(root.val);
        }
        dfs(root.right, level+1, result);
        dfs(root.left, level+1, result);
    }
}

//dfs on left

class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> result = new ArrayList<Integer>();
        dfs(root, 0, result);
        return result;
    }
    private void dfs(TreeNode root, int level, List<Integer> result){
        //base
        if(root == null) return;
        
        //logic
        if(result.size() == level){
            result.add(root.val);
        }else{
            result.set(level,root.val); // replace the values if found on right side
        }
        
        dfs(root.left, level+1, result);
        dfs(root.right, level+1, result);
    }
}

File: RightViewOfTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque


class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        
        q = collections.deque()
        
        stack = []
        
        q.append(root)
        
        
        if root == None:
            return None
        
        
        while(len(q) > 0):
            
            size = len(q)
            
            for i in range(0,size):
                
                node = q.popleft()
                if i == size -1:
                    stack.append(node.val)
                if node.left != None:
                    q.append(node.left)
                if node.right != None:
                    q.append(node.right)
        return stack
                    